token-metadata/js/src/instruction.ts

```diff
@@ -0,0 +1,166 @@
+    RemoveKey: splDiscriminate('spl_token_metadata_interface:remove_key_ix'),
+    UpdateAuthorithy: splDiscriminate('spl_token_metadata_interface:update_the_authority'),
+    Emit: splDiscriminate('spl_token_metadata_interface:emitter'),
+};
+
+// Order of keys matters
+export const SCHEMA = {
+    Initialize: { struct: { name: 'string', symbol: 'string', uri: 'string' } },
+    UpdateField: { struct: { field: 'string', value: 'string' } },
+    RemoveKey: { struct: { idempotent: 'bool', key: 'string' } },
+    UpdateAuthorithy: { array: { type: 'u8', len: 32 } },
+    Emit: { struct: { start: { option: 'u64' }, end: { option: 'u64' } } },
+};
+
+/* 
+    Initializes a TLV entry with the basic token-metadata fields.
+    
+    Assumes that the provided mint is an SPL token mint, that the metadata
+    account is allocated and assigned to the program, and that the metadata
+    account has enough lamports to cover the rent-exempt reserve.
+
+    Accounts expected by this instruction:
+      0. `[w]` Metadata
+      1. `[]` Update authority
+      2. `[]` Mint
+      3. `[s]` Mint authority
+*/
+export interface Initialize {
+    programId: PublicKey;
+    metadata: PublicKey;
+    updateAuthority: PublicKey;
+    mint: PublicKey;
+    mintAuthority: PublicKey;
+    name: string;
+    symbol: string;
+    uri: string;
+}
+
+export function createInitializeInstruction({
+    programId,
+    metadata,
+    updateAuthority,
+    mint,
+    mintAuthority,
+    name,
+    symbol,
+    uri,
+}: Initialize): TransactionInstruction {
+    const data = { name, symbol, uri };
+    return new TransactionInstruction({
+        programId,
+        keys: [
+            { isSigner: false, isWritable: true, pubkey: metadata },
+            { isSigner: false, isWritable: false, pubkey: updateAuthority },
+            { isSigner: false, isWritable: false, pubkey: mint },
+            { isSigner: true, isWritable: false, pubkey: mintAuthority },
+        ],
+        data: Buffer.concat([DISCRIMNATOR.Initialize, serialize(SCHEMA.Initialize, data)]),
+    });
+}
+
+/** If the field does not exist on the account, it will be created.
+ * If the field does exist, it will be overwritten. */
+interface UpdateFeild {
+    programId: PublicKey;
+    metadata: PublicKey;
+    updateAuthority: PublicKey;
+    field: Field;
+    value: string;
+}
+
+export function createUpdateFieldInstruction({
+    programId,
+    metadata,
+    updateAuthority,
+    field,
+    value,
+}: UpdateFeild): TransactionInstruction {
+    const data = { field, value };
+    return new TransactionInstruction({
+        programId,
+        keys: [
+            { isSigner: false, isWritable: true, pubkey: metadata },
+            { isSigner: true, isWritable: false, pubkey: updateAuthority },
+        ],
+        data: Buffer.concat([
+            DISCRIMNATOR.UpdateField,
+            Buffer.from([3]), // TODO explain this. It comes from field being typed as "Field" rather than string
```

Comment:
This appears to arise from the fact that field is of type "Field" rather than "string" in the rust code. If i convert it to a string argument, the 3 no longer appears. 
https://github.com/solana-labs/solana-program-library/blob/master/token-metadata/interface/src/instruction.rs#L255

My guess is it is some additional encoding borsh is doing, but can't figure it out. Any insights? 

I think it should be safe, because it will just be treated as a string instead - but not confident.

Reply:
Yeah you are right. Didn't realise that `Field::Key("Name".to_string())` is different from `Field::Name` (still a rust noobie)

Managed to extract the buffer when you use `Name` into a test case (but its failing).
https://github.com/solana-labs/solana-program-library/pull/5292/commits/30ed1be7bc1d5646f19eeedf859263447c803e8c

Reply:
I haven't successfully serialised this yet - but i'll take this in another thread.

But it raises the question as to what should be matched against the rust enum "Name" etc...

I think there are two solutions, using an enum or string matching

**1. Enum**
Define a typescript enum, and encode as the enum only if the enum is provided
```
enum RequiredField {
  Name = 0 ,
  Symbol = 1,
  Uri = 2,
}
type Field = RequiredField | "string"
```

**2. string matching**
```
   if(field == "Name" || field == "name" ) {
    // encode as enum Field::name
  } else {
     // encode as user defined key
  }

```

Solution 1 more closely matches the rust implementation and allows there to both keys `Field::Name` and `Field::Key("Name".to_string())` (which i think can co-exist based on my limited understanding?). However Enums is a typescript feature only, and any javascript users would be out of like. I think they would need to pass in 0 as the field to get the correct behaviour'



I'm more inclined to go with Solution 1 - if its really true that `Field::Name` and `Field::Key("Name".to_string())` can co-exist